sycl/doc/design/OptionalDeviceFeatures.md

```diff
@@ -913,211 +913,253 @@ aspects in the array.
+device names might change too in order to align with the naming style used by
+other offload architectures.  I presume we want to use the same strings for
+both `--offload-arch` and for the `-target` option to `clang-offload-wrapper`.
+Therefore, the target names presented above might change.
+
+**TODO**: The handling of the SPIR-V offload bundle requires some more thought.
+In the current design, there are two different types of SPIR-V modules, which
+are not exactly the same.
+
+If the user compiles with `-fsycl-targets=spir64` (which is the default), the
+device compiler is invoked with `-triple spir64-unknown-unknown`, and the
+generated IR is converted to SPIR-V.  This type of SPIR-V module is wrapped
+into the host executable via `clang-offload-wrapper -target=spir64` (i.e. the
+offload bundle has type "spir64").
+
+However, if the user AOT compiles for an Intel GPU with
+`-fsycl-targets=spir64_gen`, the device compiler is invoked with
+`-triple spir64_gen-unknown-unknown`, which is a different triple from the case
+above.  This IR is also converted to SPIR-V and then sent to the `ocloc` tool.
+The `ocloc` tool compiles the SPIR-V to native code but also embeds the SPIR-V
+into the ocloc output.  This SPIR-V is used by the device driver if the
+application is run on a GPU device that does not match any of the AOT targets.
+
+In the new design, `ocloc` does not embed SPIR-V in its output because we
+invoke it with `-exclude_ir`.  Instead, DPC++ is responsible for embedding the
+SPIR-V in the host executable, but should that SPIR-V be generated with the
+triple `spir64-unknown-unknown` or the triple `spir64_gen-unknown-unknown`?
+We could generate both, but that seems wasteful.  If we generate the SPIR-V
+with `spir64-unknown-unknown`, does that mean we need to run the device
+compiler twice, once with `spir64_gen-unknown-unknown` (to generate SPIR-V that
+is fed to `ocloc`) and once with `spir64-unknown-unknown` (to generate SPIR-V
+that is embedded in the host executable)?  This also seems wasteful because
+it requires running the device compiler twice whenever the user AOT compiles
+for an Intel GPU.
+
+An ideal solution would be to always generate SPIR-V using the triple
+`spir64-unknown-unknown`, and use this SPIR-V as input to both `ocloc` and
+`opencl-aot`.  This would allow us to run the device compiler just once even if
+the user AOT compiles for both Intel GPU and CPU.  However, this means that the
+SPIR-V is generated using a different triple than we use today.  It's unclear
+what ramification this has.
+
+
 ### Changes to the DPC++ runtime
 
-The DPC++ runtime must be changed to check if a kernel uses any optional
-features that the device does not support.  If this happens, the runtime must
-raise a synchronous `errc::kernel_not_supported` exception.
+#### Identifying a compatible offload bundle
+
+Since the user may AOT compile for several different offload targets, the
```

Comment:
Ideally, this should be moved down to PI/UR layer, because it is too backend-specific.

There are different approaches possible:
- an API which checks if a device image is compatible with particular device/arch
- an API which accepts a set of device images and returns an index of a "best" for a specified device/target/arch

We have though about adding the second API into our SW stack to implement AOT -> JIT fallback a while ago. However,  if we also factor in some extra features that may affect selection of a device image, like presence or manipulations with specialization constants, then the first API may be a better fit for our needs.

Reply:
Yes, I agree that we need some new PI/UR API here.  I guess I was trying to avoid the details of where in the runtime we implement the check.